source/Halibut/Transport/Streams/ReadIntoMemoryBufferStream.cs

```diff
@@ -0,0 +1,126 @@
+        {
+            memoryStream = new MemoryStream();
+            this.sourceStream = sourceStream;
+            this.readIntoMemoryLimitBytes = readIntoMemoryLimitBytes;
+            this.onDispose = onDispose;
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            base.Dispose(disposing);
+
+            if (disposing)
+            {
+                memoryStream.Dispose();
+
+                if (onDispose == OnDispose.DisposeInputStream)
+                {
+                    sourceStream.Dispose();
+                }
+            }
+        }
+
+        bool ShouldReadFromMemoryStream
+        {
+            get
+            {
+                if (BytesReadIntoMemory == 0) return false;
+
+                return !limitReached || memoryStream.Position < BytesReadIntoMemory;
+            }
+        }
+
+        public long BytesReadIntoMemory => memoryStream.Length;
+
+        public override bool CanRead => sourceStream.CanRead;
+        public override bool CanWrite => false;
+        public override bool CanSeek => false;
+        public override bool CanTimeout => false;
+
+        public override long Length => throw new NotSupportedException();
+
+        public override long Position
+        {
+            get => throw new NotSupportedException();
+            set => throw new NotSupportedException();
+        }
+
+        public override int ReadTimeout
+        {
+            get => sourceStream.ReadTimeout;
+            set => throw new NotSupportedException();
+        }
+
+        public override int WriteTimeout
+        {
+            get => sourceStream.WriteTimeout;
+            set => throw new NotSupportedException();
+        }
+
+        public override void Flush() => throw new NotSupportedException();
+
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            if (ShouldReadFromMemoryStream)
+            {
+                return memoryStream.Read(buffer, offset, count);
+            }
+
+            return sourceStream.Read(buffer, offset, count);
+        }
+
+        public override async Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            if (ShouldReadFromMemoryStream)
+            {
+                return await memoryStream.ReadAsync(buffer, offset, count, cancellationToken);
+            }
+
+            return await sourceStream.ReadAsync(buffer, offset, count, cancellationToken);
+        }
+
+        public override long Seek(long offset, SeekOrigin origin) => throw new NotSupportedException();
+
+        public override void SetLength(long value) => throw new NotSupportedException();
+        public override void Write(byte[] buffer, int offset, int count) => throw new NotSupportedException();
+
+        public async Task BufferIntoMemoryFromSourceStreamUntilLimitReached(CancellationToken cancellationToken)
+        {
+            var buffer = new byte[81920];
```

Comment:
why `81920`?

Reply:
This is the value of `Stream._DefaultCopyBufferSize` (which is not public 😑)

Reply:
This might not be what we want since now we alloc 81k for each in flight request we are waiting for, bring it down to `8192` or what the buffered rewindable stream has this set to which is under HalibutLimits.

The rewindable stream will actually only fetch up to its buffer size, so it wont be any slower.

Reply:
Wait no actually the rewindable buffer stream is on the compressed bytes but this will be looking at uncompressed bytes.

I think the memory concerns still hold, so dropping this buffer size down is probably going to reduce memory pressure without much or any performance.

Reply:
I moved this down to 8192 here, as well as in the `WriteIntoMemoryBufferStream`